### DISCUSSION TITLE
Delegate패턴이 무엇일까요?

### DISCUSSION THREAD

**Discussion by JUNY0110** [Samsamhada/MacC-Team-Samsamhada#112] [General]
- 제가 이해하고 사용한 Delegate와 Protocol에 대한 내용입니다.
  - Delegate는 객체가 자신의 책임을 다른 객체에게 위임(delegate)하는 디자인 패턴을 말합니다. 먼저 이렇게만 이해하고 넘어가고,
  - TableViewDelegate, TableViewDataSource안에는 속이 빈 내부구현코드(protocol)이 있습니다.
  - <img width="897" alt="스크린샷 2022-11-10 오후 4 03 59" src="https://user-images.githubusercontent.com/98405970/201022684-36f1eb71-f29d-4ab8-b646-bcb0d4930545.png">
  - TableView에서 사용할 기능에 대해 추상화된 추상클래스(Protocol)를 필요할 때 꺼내오는 것이며, 구체화된 코드를 tableView.delegate = self를 통해 ViewController에서 위임받아 참조(인식)하도록 만드는 거라고 추측됩니다.
  -  하지만 Class참조의 경우, Strong(강한) 참조로 연결되어 있기 때문에 참조가 쉽게 끊기지 않아 Memory Leak(메모리 누수) 문제가 발생할 가능성이 높습니다. 따라서 delegate를 weak(약한) 참조로 선언하여 메모리 누수 문제를 예방할 수 있습니다. 또한 Protocol을 AnyObject로 하는 이유는, Struct(구조체) 선언 시 참조가 아닌 복제(?)가 되기 때문에 참조되지 않아 감지할 수 없기 때문입니다.
  - 저는 delegate를 선언한 곳은 `변화가 일어날 곳`이며, delegate를 받는 곳은 `변화를 인지하는 곳`이라고 생각했습니다.
  - <img width="897" alt="스크린샷 2022-11-10 오후 4 51 19" src="https://user-images.githubusercontent.com/98405970/201032351-8def0a1e-62b3-448a-a196-8d8dde135216.png">
  - 그래서, UITableViewCell에서 객체의 행위(DatePicker)를 전달할 필요가 있고, 변화(DatePicker 선택날짜 변화)를 ViewController에서 인식하도록 하기 위해 ViewController에서 delegate = self 를 작성했습니다.
  - Protocol은 delegate를 통해 `변화를 인지하는 곳`에서 입맛에 맞게 구체화하는 작업을 extension에서 구현하게 됩니다.
- 설명이 잘못된 부분이 있다면 수정해주시면 감사하겠습니다. 글로 정리하려고 하니까 잘 정리가 안 되네요; 해당 주제에 대한 자유로운 토론 환영입니다.

### 참고자료
- [Delegate패턴 이해1](https://velog.io/@nala/iOS-Delegate-패턴을-이해해보자)
- [Delegate패턴 이해2](https://june0122.github.io/2021/08/21/design-pattern-delegate/)
- [Delegate- class 사용에 대한 내용](https://velog.io/@marisol/Delegate-retain이-될까)
- [Delegate 순환참조에 따른 Memory Leak 발생 예방](https://goodmorningcody.wordpress.com/2016/11/24/델리게이트-패턴-사용-시-순환참조에-의한-메모리릭/)

**Comment by ikanic** [comment]
> Protocol을 AnyObject로 하는 이유는, Struct(구조체) 선언 시 참조가 아닌 복제(?)가 되기 때문에 참조되지 않아 감지할 수 없기 때문입니다.

해당 부분에 대해 첨언해보자면, Class는 기본적으로 얕은 복사(Shallow Copy)를 하기 때문에, 데이터를 복사해도 인스턴스 메모리가 새로 생성되는것이 아닌 값의 주소를 복사하여 같은 값을 가르키게 됩니다.
반면에, Struct의 경우 기본적으로 깊은 복사(Deep Copy)를 하기 때문에, 데이터를 그대로 복사하여 복사된 두 데이터 객체는 각각의 메모리를 가지게 됩니다.

해당 부분은 `Shallow Copy`와 `Deep Copy`라는 키워드로 찾아보면 좋을 것 같으며, 아래에 레퍼런스 하나 첨부합니다.

### 레퍼런스
- [Just Do iOS - Deep copy & Shallow copy](https://green1229.tistory.com/126)

**Comment by ikanic** [comment]
> TableView에서 사용할 기능에 대해 추상화된 추상클래스(Protocol)를 필요할 때 꺼내오는 것이며, 구체화된 코드를 tableView.delegate = self를 통해 ViewController에서 위임받아 참조(인식)하도록 만드는 거라고 추측됩니다.

해당 부분에서 한가지 말씀드려보자면, Protocol과 추상클래스(Abstract Class)는 비슷하지만 다른 용어입니다.
자세한 내용은 비슷하지만 서로 다른 `추상클래스(Abstract Class)`, `Interface`, `Protocol`이라는 키워드로 찾아보시면 더 좋을 것 같습니다.

사실 Swift에서 추상클래스라는 용어를 사용하는지는 잘 모르겠습니다만, 일반적으로 추상 클래스는 추상화된 클래스로, 기본 동작을 구현한 인스턴스 메서드를 가질 수 있습니다. 그에 반해 인스턴스는 메서드가 무조건 추상 메서드여야 한다는 조건이 있습니다.
또, 인터페이스는 추상화된 메서드를 무조건 구현해야한다는 조건이 있지만, 프로토콜은 메서드에도 옵셔널 키워드를 사용할 수 있어 추상화된 메서드를 선택적으로 구현할 수 있습니다.

자세한 내용은 직접 찾아보시면 좋을 것 같고, 아래에 참고할 만한 간단한 레퍼런스 첨부하겠습니다.
- [wan088.log - 추상클래스 vs 인터페이스 vs 프로토콜](https://velog.io/@wan088/%EC%B6%94%EC%83%81-%ED%81%B4%EB%9E%98%EC%8A%A4-vs-%EC%9D%B8%ED%84%B0%ED%8E%98%EC%9D%B4%EC%8A%A4-vs-%ED%94%84%EB%A1%9C%ED%86%A0%EC%BD%9C-jmjztgi1kp)

추가로 Swift에서 Protocol을 추상클래스나 인터페이스처럼 사용하는 방법에 대한 글이 있길래 함께 첨부합니다.
- [Diligent Developer - #10 Swift - Protocol as Interface](https://ally10.tistory.com/39)
